File: src/types/service.ts
/**
 * @fileoverview Type definitions for services in the dependency mapping system.
 *
 * Defines the core Service interface and related types used throughout
 * the application for representing microservices and their metadata.
 *
 * @module types/service
 */

/**
 * Supported service type categories.
 *
 * Each type represents a common architectural component:
 * - `api`: REST or GraphQL API service
 * - `database`: Persistent data storage (PostgreSQL, MongoDB, etc.)
 * - `cache`: In-memory cache (Redis, Memcached)
 * - `queue`: Message queue (RabbitMQ, Kafka, SQS)
 * - `gateway`: API gateway or load balancer
 * - `frontend`: Web or mobile frontend application
 * - `backend`: Backend processing service
 * - `external`: Third-party or external service
 *
 * Also accepts arbitrary strings for custom types.
 */
export type ServiceType =
  | "api"
  | "database"
  | "cache"
  | "queue"
  | "gateway"
  | "frontend"
  | "backend"
  | "external"
  | "identity-provider"
  | string;

/**
 * Health status indicators for services.
 *
 * - `healthy`: Service is operating normally
 * - `degraded`: Service is operational but experiencing issues
 * - `unhealthy`: Service is not functioning correctly
 * - `unknown`: Health status cannot be determined
 * - `deprecated`: Service is scheduled for removal
 */
export type ServiceStatus =
  | "healthy"
  | "degraded"
  | "unhealthy"
  | "unknown"
  | "deprecated";

/**
 * Represents a service/microservice in the dependency graph.
 *
 * This is the core data structure for services, containing all metadata
 * needed for visualization, filtering, and documentation.
 *
 * @property id - Unique identifier (lowercase alphanumeric with hyphens)
 * @property name - Human-readable display name
 * @property serviceType - Category of service (api, database, etc.)
 * @property status - Current health status
 * @property description - Optional detailed description
 * @property version - Optional version string (e.g., "1.0.0")
 * @property owner - Optional individual owner identifier
 * @property team - Optional team responsible for the service
 * @property tags - Array of searchable tags for categorization
 * @property metadata - Key-value pairs for custom properties
 *
 * @example
 * ```typescript
 * const service: Service = {
 *   id: "user-api",
 *   name: "User API",
 *   serviceType: "api",
 *   status: "healthy",
 *   description: "Handles user authentication and profiles",
 *   version: "2.1.0",
 *   owner: "john.doe",
 *   team: "Platform",
 *   tags: ["auth", "critical", "public-facing"],
 *   metadata: { port: 8080, replicas: 3 }
 * };
 * ```
 */
export interface Service {
  id: string;
  name: string;
  serviceType: ServiceType;
  status: ServiceStatus;
  description?: string;
  version?: string;
  owner?: string;
  team?: string;
  tags: string[];
  metadata: Record<string, unknown>;
}

File: src-tauri/src/models/service.rs
//! Service data model definitions.
//!
//! This module defines the core `Service` type and its associated enums
//! for representing services in the dependency graph.

use serde::{Deserialize, Serialize};
use std::collections::HashMap;

/// The type/category of a service in the architecture.
///
/// Used to classify services for filtering and visual differentiation
/// in the dependency graph.
///
/// # Variants
///
/// * `Api` - REST API or GraphQL endpoint services
/// * `Database` - Database services (PostgreSQL, MySQL, MongoDB, etc.)
/// * `Cache` - Caching services (Redis, Memcached, etc.)
/// * `Queue` - Message queue services (RabbitMQ, Kafka, SQS, etc.)
/// * `Gateway` - API gateways or load balancers
/// * `Frontend` - Client-facing web applications
/// * `Backend` - Backend microservices or monolithic applications
/// * `External` - Third-party or external services
/// * `Custom(String)` - Custom type for extensibility (catch-all)
///
/// # Serialization
///
/// Types are serialized as snake_case strings (e.g., `ServiceType::Api` → `"api"`).
/// Custom types are serialized directly as their string value.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ServiceType {
    Api,
    Database,
    Cache,
    Queue,
    Gateway,
    Frontend,
    Backend,
    External,
    #[serde(untagged)]
    Custom(String),
}

impl Default for ServiceType {
    /// Returns the default service type: `Backend`.
    fn default() -> Self {
        ServiceType::Backend
    }
}

/// The operational status of a service.
///
/// Used to indicate the current health state of a service for
/// monitoring and visualization purposes.
///
/// # Variants
///
/// * `Healthy` - Service is operating normally
/// * `Degraded` - Service is running but with reduced capacity or issues
/// * `Unhealthy` - Service is experiencing significant problems
/// * `Unknown` - Service status cannot be determined
/// * `Deprecated` - Service is marked for removal and should not be used
///
/// # Serialization
///
/// Statuses are serialized as snake_case strings.
#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum ServiceStatus {
    Healthy,
    Degraded,
    Unhealthy,
    Unknown,
    Deprecated,
}

impl Default for ServiceStatus {
    /// Returns the default service status: `Unknown`.
    fn default() -> Self {
        ServiceStatus::Unknown
    }
}

/// Represents a service in the dependency graph.
///
/// A service is any distinct component in the architecture that can have
/// dependencies on or from other services. This is the core data model
/// used throughout the application.
///
/// # Required Fields
///
/// * `id` - Unique identifier for the service (used as filename)
/// * `name` - Human-readable display name
///
/// # Optional Fields
///
/// All other fields have sensible defaults and are optional during creation.
///
/// # Serialization
///
/// Fields are serialized to camelCase for JavaScript/TypeScript compatibility:
/// - `service_type` → `serviceType`
/// - `description` → `description`
///
/// # Example JSON
///
/// ```json
/// {
///   "id": "user-service",
///   "name": "User Service",
///   "serviceType": "api",
///   "status": "healthy",
///   "description": "Handles user authentication and profile management",
///   "version": "2.1.0",
///   "owner": "auth-team@company.com",
///   "team": "Authentication Team",
///   "tags": ["auth", "users", "core"],
///   "metadata": {
///     "repository": "https://github.com/company/user-service",
///     "port": 8080
///   }
/// }
/// ```
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Service {
    /// Unique identifier for the service. Used as the filename and for references.
    pub id: String,
    /// Human-readable display name for the service.
    pub name: String,
    /// The category/type of service (defaults to Backend).
    #[serde(default)]
    pub service_type: ServiceType,
    /// Current operational status (defaults to Unknown).
    #[serde(default)]
    pub status: ServiceStatus,
    /// Optional detailed description of the service's purpose.
    #[serde(default)]
    pub description: Option<String>,
    /// Optional version string (e.g., "1.2.3" or "v2.0.0").
    #[serde(default)]
    pub version: Option<String>,
    /// Optional owner email or identifier.
    #[serde(default)]
    pub owner: Option<String>,
    /// Optional team name responsible for the service.
    #[serde(default)]
    pub team: Option<String>,
    /// Tags for filtering and categorization.
    #[serde(default)]
    pub tags: Vec<String>,
    /// Arbitrary key-value metadata for extensibility.
    #[serde(default)]
    pub metadata: HashMap<String, serde_json::Value>,
}

impl Service {
    /// Checks if the service matches a search query.
    ///
    /// Performs a case-insensitive substring search across multiple fields
    /// of the service. Returns true if the query is found in any of:
    /// - `name`
    /// - `id`
    /// - `description`
    /// - `owner`
    /// - `team`
    /// - Any tag in `tags`
    ///
    /// # Arguments
    ///
    /// * `query` - The search string to match (case-insensitive)
    ///
    /// # Returns
    ///
    /// `true` if the query matches any searchable field, `false` otherwise.
    ///
    /// # Examples
    ///
    /// ```rust
    /// let service = Service {
    ///     id: "user-api".to_string(),
    ///     name: "User API Service".to_string(),
    ///     tags: vec!["auth".to_string(), "api".to_string()],
    ///     // ... other fields
    /// };
    ///
    /// assert!(service.matches_search("user"));   // matches id and name
    /// assert!(service.matches_search("AUTH"));   // matches tag (case-insensitive)
    /// assert!(!service.matches_search("orders")); // no match
    /// ```
    pub fn matches_search(&self, query: &str) -> bool {
        let query_lower = query.to_lowercase();

        self.name.to_lowercase().contains(&query_lower)
            || self.id.to_lowercase().contains(&query_lower)
            || self.description
                .as_ref()
                .map(|d| d.to_lowercase().contains(&query_lower))
                .unwrap_or(false)
            || self.owner
                .as_ref()
                .map(|o| o.to_lowercase().contains(&query_lower))
                .unwrap_or(false)
            || self.team
                .as_ref()
                .map(|t| t.to_lowercase().contains(&query_lower))
                .unwrap_or(false)
            || self.tags.iter().any(|tag| tag.to_lowercase().contains(&query_lower))
    }
}
